src/mlstacks/utils/terraform_utils.py

```diff
@@ -0,0 +1,621 @@
+
+    # run Terraform
+    tfr = TerraformRunner(tf_recipe_path)
+    if not tf_previously_initialized(tf_recipe_path):
+        tf_client_init(tfr.client, provider=stack.provider, debug=debug_mode)
+
+        # write a file with name `IGNORE_ME` to the Terraform recipe directory
+        # to prevent Terraform from initializing the recipe
+        Path(f"{tf_recipe_path}/{MLSTACKS_INITIALIZATION_FILE_FLAG}").touch()
+
+    # TODO: confirm the logging of progress doesn't require user input still
+    # confirm the plan
+    # log what's being deployed
+    # spinner to state that Terraform is running
+    # output the outputs at the end (or in CLI?)
+    tf_client_apply(
+        client=tfr.client,
+        tf_vars=tf_vars,
+        debug=debug_mode,
+        provider=stack.provider,
+    )
+
+
+def destroy_stack(stack_path: str, debug_mode: bool = False) -> None:
+    """Destroy stack.
+
+    Args:
+        stack_path: The path to the stack.
+        debug_mode: Whether to run in debug mode.
+    """
+    stack = load_stack_yaml(stack_path)
+    tf_vars = parse_tf_vars(stack)
+
+    tf_recipe_path = f"{CONFIG_DIR}/terraform/{stack.provider}-modular"
+
+    tfr = TerraformRunner(tf_recipe_path)
+
+    if not tf_previously_initialized(tf_recipe_path):
+        tf_client_init(tfr.client, provider=stack.provider, debug=debug_mode)
+
+        # write a file with name `IGNORE_ME` to the Terraform recipe directory
+        # to prevent Terraform from initializing the recipe
+        Path(f"{tf_recipe_path}/{MLSTACKS_INITIALIZATION_FILE_FLAG}").touch()
+
+    tf_client_destroy(tfr.client, tf_vars, debug_mode)
+
+
+def get_stack_outputs(
+    stack_path: str,
+    output_key: Optional[str] = None,
+    debug_mode: bool = False,
+) -> Dict[str, str]:
+    """Get stack outputs.
+
+    Args:
+        stack_path: The path to the stack.
+        output_key: The output key.
+
+    Returns:
+        The stack outputs.
+
+    Raises:
+        RuntimeError: If Terraform has not been initialized.
+    """
+    stack = load_stack_yaml(stack_path)
+    tf_recipe_path = f"{CONFIG_DIR}/terraform/{stack.provider}-modular"
+    state_tf_path = f"{tf_recipe_path}/terraform.tfstate"
+
+    tfr = TerraformRunner(tf_recipe_path)
+    if not tf_previously_initialized(tf_recipe_path):
+        raise RuntimeError(
+            "Terraform has not been initialized so there are no "
+            "outputs to show."
+        )
+
+    # TODO: add debug mode functionality
+    # TODO: extract out into separate helper method
+    if output_key:
+        full_outputs = tfr.client.output(
+            output_key, full_value=True, state=state_tf_path
+        )
+        return {output_key: full_outputs}
+    else:
+        full_outputs = tfr.client.output(full_value=True, state=state_tf_path)
+        return {k: v["value"] for k, v in full_outputs.items() if v["value"]}
+
+
+def infracost_installed(f: Callable[..., Any]) -> Callable[..., Any]:
```

Comment:
personally not a fan of having these sort of things defined in a decorator, calling a normal `check_infracost_installed` function is much more readable and intuitive IMO